### PR TITLE
added logprobs (`n_probs`)

### PR DESCRIPTION
**Pull request by janpf**
As discussed on discord I implemented the feature. It just passes through the probs from the llamacpp server. Sorry, first time writing Go, might have missed something.

https://discord.com/channels/1128867683291627614/1128867684130508875/1187028494228664340

### CODE REVIEW

**Comment by janpf** [comment]
Just noticed the entire llama.go file got rewritten in the meantime :/ maybe the person refactoring llama.go could give a hint at where to implement it now? 🥺👉🏻👈🏻 I took a quick look in the online merge editor but it’s not trivial.

Happy Holidays!

**Comment by janpf** [comment]
```json
janpf@p085info010013 ~> curl http://localhost:11434/api/generate -d '{"model":"llama2", "temperature":0, "prompt": "Give me your probabilities", "options": {"num_predict": 2, "n_probs":2}}' | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1722    0  1603  100   119   4734    351 --:--:-- --:--:-- --:--:--  5094
{
  "model": "llama2",
  "created_at": "2024-01-08T12:11:42.615133Z",
  "response": "\n",
  "completion_probabilities": [
    {
      "content": "\n",
      "probs": [
        {
          "prob": 0.5468829274177551,
          "tok_str": "As"
        },
        {
          "prob": 0.38397082686424255,
          "tok_str": "\n"
        }
      ]
    }
  ],
  "done": false
}
{
  "model": "llama2",
  "created_at": "2024-01-08T12:11:42.632652Z",
  "response": "As",
  "completion_probabilities": [
    {
      "content": "As",
      "probs": [
        {
          "prob": 0.6612563133239746,
          "tok_str": "As"
        },
        {
          "prob": 0.3387437164783478,
          "tok_str": "I"
        }
      ]
    }
  ],
  "done": false
}
{
  "model": "llama2",
  "created_at": "2024-01-08T12:11:42.650095Z",
  "response": " a",
  "completion_probabilities": [
    {
      "content": " a",
      "probs": [
        {
          "prob": 0.9904254078865051,
          "tok_str": " a"
        },
        {
          "prob": 0.009574598632752895,
          "tok_str": " an"
        }
      ]
    }
  ],
  "done": false
}
{
  "model": "llama2",
  "created_at": "2024-01-08T12:11:42.667608Z",
  "response": " responsible",
  "completion_probabilities": [
    {
      "content": " responsible",
      "probs": [
        {
          "prob": 0.9597197771072388,
          "tok_str": " responsible"
        },
        {
          "prob": 0.04028024524450302,
          "tok_str": " neutral"
        }
      ]
    }
  ],
  "done": false
}
{
  "model": "llama2",
  "created_at": "2024-01-08T12:11:42.685076Z",
  "response": " A",
  "completion_probabilities": [
    {
      "content": " A",
      "probs": [
        {
          "prob": 0.9497269988059998,
          "tok_str": " A"
        },
        {
          "prob": 0.050273049622774124,
          "tok_str": " and"
        }
      ]
    }
  ],
  "done": false
}
{
  "model": "llama2",
  "created_at": "2024-01-08T12:11:42.685199Z",
  "response": "",
  "done": true,
  "context": [
    518,
    25580,
    29962,
    3532,
    14816,
    29903,
    29958,
    5299,
    829,
    14816,
    29903,
    6778,
    13,
    13,
    29954,
    573,
    592,
    596,
    2070,
    11614,
    518,
    29914,
    25580,
    29962,
    13,
    13,
    2887,
    263,
    14040,
    319
  ],
  "total_duration": 337091792,
  "load_duration": 3560750,
  "prompt_eval_count": 11,
  "prompt_eval_duration": 267993000,
  "eval_count": 4,
  "eval_duration": 52417000
}
```

Two issues:
- it does not really respect `num_predict` (which is an issue that is also existing for me on the main branch, and I don't see how the content of this PR could be responsible for this) and
- for the first token commonly the token with the smaller score is selected (again an issue that is unlikely to stem from my code, as I'm just passing through the logits from the underlying server)

**Comment by briancleland** [comment]
Hi, any indication of when this might be merged?

**Comment by caj-larsson** [comment]
Is there any problem with merging this? I can probably help out if there is something more needs doing. This is blocking my usecase and I would love to get away from using llama.cpp directly.

**Comment by briancleland** [comment]
Just bumping this again to see if it can be merged soon.

**Comment by iurimatias** [comment]
hi @bmizerany , can this PR be merged? It's a very useful feature missing from Ollama currently

**Comment by bmizerany** [comment]
Thank you for your PR. We're working on things this could break or might be broken by, so we're going to wait until we've had more time to make sure this lines up with those. Thank you for your patience and support!

**Comment by caj-larsson** [comment]
> Thank you for your PR. We're working on things this could break or might be broken by, so we're going to wait until we've had more time to make sure this lines up with those. Thank you for your patience and support!

Sorry for my ignorance but this is just exposing the core functionality of the library this project is a wrapper for.

Mind linking that conversation? I'm happy to help

**Comment by briancleland** [comment]
> Thank you for your PR. We're working on things this could break or might be broken by, so we're going to wait until we've had more time to make sure this lines up with those. Thank you for your patience and support!

Hi @bmizerany, thanks for the update. Do you have an approximate timescale for this work?

**Comment by bmizerany** [comment]
@briancleland We do not currently have a timeline to support this.

**Comment by briancleland** [comment]
> We're working on things this could break or might be broken by, so we're going to wait until we've had more time to make sure this lines up with those.

@bmizerany On reflection, I think my last question was not very clear. What I meant was, do you know when the work referred to above might be completed?

**Comment by caj-larsson** [comment]
@briancleland @janpf Anyone else feel like it might be time for a hard fork with emphasis on full LLM-capabilities exposed in APIs?

**Comment by briancleland** [comment]
@caj-larsson I would hope that wouldn't be necessary, but the unhelpful response to this PR is a bit puzzling.

**Comment by boxabirds** [comment]
Ollama is an amazing piece of tech! I have used it extensively for quite some time -- huge kudos to the team! I'm also interested in logprobs being exposed for two reasons
1. To maintain compatibility with the OpenAI library [which does include it](https://platform.openai.com/docs/api-reference/chat/create#chat-create-logprobs)
2. To provide better insight into the quality of the result.

Very much looking forward to seeing this integrated! Happy to help but I would expect this is not a starter-contributor ticket.

**Comment by MaxOmlor** [comment]
Hello there,
I'm also really looking forward to using logprobs for my project.
Any news on when this feature will be integrated?
Greetings

**Comment by eliot-christon** [comment]
> Hello there, I'm also really looking forward to using logprobs for my project. Any news on when this feature will be integrated? Greetings

Hello,
Same for evaluating models with ollama through [lm-evaluation-harness](https://github.com/EleutherAI/lm-evaluation-harness).
Very much looking forward to seeing this integrated!

**Comment by briancleland** [comment]
@jmorganca Any thoughts on how to progress this?

**Comment by The-Inscrutable-X** [comment]
I would like to push for this feature as well. It's critical for confidence evaluation using both logprobs and entropy.

**Comment by wiiiktor** [comment]
Top layer logits would be very useful for potential transfer learning, for example with additional layers.

**Comment by TensorTemplar** [comment]
> @briancleland @janpf Anyone else feel like it might be time for a hard fork with emphasis on full LLM-capabilities exposed in APIs?

I am like 99.99% with you here, where the missing 0.01% is not knowing any golang at all. 
It seems the lesser effort would be figuring how to expose capabilities already in llama.cpp than to deal with managing the model ecosystem without ollama.

**Comment by OriginalGoku** [comment]
I can not believe that implementing such simple and useful feature has been pending for over 7 months with no clear path forward! 
This is more like a simple print statement! I offer my help to fixing this issue if I know whats pending.

**Comment by drdsgvo** [comment]
> I can not believe that implementing such simple and useful feature has been pending for over 7 months with no clear path forward! This is more like a simple print statement! I offer my help to fixing this issue if I know whats pending.

I also do not understand why ollama does not support logits. It is a fundamental aspect of an LLM. A lot of serious algorithms, that enhance the quality of the LLM output, rely on logits.

Please, Ollama team, do support logits ASAP. Thank you!

**Comment by Bruno-TT** [comment]
Get it together Ollama team. This has been sitting unmerged for 7 months.

**Comment by josiahbryan** [comment]
Any news here? Why is this taking so long?

**Comment by magic-YuanTian** [comment]
You should be aware that this is a very important feature. Missing it will cause you to lose a significant amount of market.

**Comment by michaelgfeldman** [comment]
Please add. This is a very important feature.

**Review by OriginalGoku — APPROVED** [review]

**Comment by avsolatorio** [comment]
+1 for this feature! It’s limiting the features that can be done with ollama if there is no possibility of getting the logprobs back. Hope this gets merged very soon! 🤞🏽

**Comment by ToluClassics** [comment]
+1 for this feature

**Comment by Luxter77** [comment]
+1 would love to have this one feature

**Comment by briancleland** [comment]
Would it be possible for one of the maintainers to give an idea of when this might be implemented?

**Comment by ericcurtin** [comment]
We have an Ollama-compatible client built here (and it can also pull huggingface models and models from OCI registries):

https://github.com/containers/ramalama

feel free to contribute similar features here, they won't be ignored...

**Comment by briancleland** [comment]
> We have an Ollama-compatible client built here (and it can also pull huggingface models and models from OCI registries):
> 
> https://github.com/containers/ramalama
> 
> feel free to contribute similar features here, they won't be ignored...

Thanks, will check it out!

**Comment by drdsgvo** [comment]
> We have an Ollama-compatible client built here (and it can also pull huggingface models and models from OCI registries):
> 
> https://github.com/containers/ramalama
> 
> feel free to contribute similar features here, they won't be ignored...
Great to hear and see! Ollama sucks as they are not able to implement a very important features. They even do not see the importance, which is the saddest part. Shame on them.

**Comment by to-sora** [comment]
please add

**Comment by Cy-Fi** [comment]
+1

**Comment by codelion** [comment]
We now support logprobs directly in our local inference server - https://github.com/codelion/optillm?tab=readme-ov-file#local-inference-server

**Comment by sammcj** [comment]
I'd be happy to look at how this PR could be updated/re-implemented to suit the current Ollama codebase with it's new runners/server/cgo implementation but want to wait until after #6279 is merged so I'm not tripping up over myself.

**Comment by ParthSareen** [comment]
Hey everyone - really sorry for the long delay here. This is something on top of mind that I want to get out in Jan or so. Hopefully will get this out soon! If anyone is interested in taking a pass at it - we can get it to the finish line together :) cc @sammcj @janpf

**Comment by OriginalGoku** [comment]
@ParthSareen Great to hear that you guys are taking this seriously.
How can we help?

**Comment by p11o** [comment]
This is critical to CoT configurations. Anything I can do to help with this?

**Comment by Lendle-King** [comment]
+1 Definitely need it!

**Comment by ParthSareen** [comment]
Will be tracking progress in: https://github.com/ollama/ollama/issues/2415
